substratools/task_resources.py

```diff
@@ -2,72 +2,103 @@
+                item[_RESOURCE_ID], {_RESOURCE_VALUE: [], _RESOURCE_MULTIPLE: item[_RESOURCE_MULTIPLE]}
+            )
+            self._values[item[_RESOURCE_ID]][_RESOURCE_VALUE].append(item[_RESOURCE_VALUE])
 
-    def get_value(self, key: str) -> str:
-        """Return a value corresponding to the given key.
-        It will raise if there is no input for given key or if there are multiple values"""
-        val = self._values[key]
-        if len(val) > 1:
-            raise exceptions.InvalidInputOutputsError("there are more than one path")
+        self.opener_path = self.get_value(TASK_IO_OPENER)
+        self.input_data_folder_paths = self.get_value(TASK_IO_DATASAMPLES)
+        self.chainkeys_path = self.get_value(TASK_IO_CHAINKEYS)
 
-        return val[0]
+    def get_value(self, key: str) -> Optional[Union[List[str], str]]:
+        """Returns the value for a given key. Won't raise if there is no matching resource.
+        Will raise if there is a mismatch between the given multiplicity and the number of returned
+        elements.
 
-    def get_values(self, key: str) -> List[str]:
-        """get_values will return the list of str corresponding to the given key.
-        It will raise if key does not exist.
+        If multiple is True, will return a list else will return a single value
         """
-        return self._values[key]
-
-    def get_optional_value(self, key: str) -> Optional[str]:
-        """Return value for given key, won't raise if there is no matching resource.
-        Will raise if there are more than one value."""
         if key not in self._values:
             return None
-        val = self._values[key]
 
-        if len(val) > 1:
-            raise exceptions.InvalidInputOutputsError("there are more than one path")
+        val = self._values[key][_RESOURCE_VALUE]
+        multiple = self._values[key][_RESOURCE_MULTIPLE]
+
+        if not multiple and len(val) > 1:
+            raise exceptions.InvalidInputOutputsError(
+                f"There is more than one path for the non multiple resource {key}"
+            )
+
+        if multiple:
+            return val
 
         return val[0]
 
-    def get_optional_values(self, key: str) -> Optional[List[str]]:
-        """Return values for given key, won't raise if there is no matching resource."""
-        return self._values.get(key)
+    @property
+    def formatted_dynamic_resources(self) -> Union[List[str], str]:
+        """Returns all the resources (except the datasamples, the opener and the chainkeys_path under the user format:
+        A dict where each input is an element where
+            - the key is the user identifier
+            - the value is a list of Path for multiple resources and a Path for non multiple resources
+        """
+
+        return {
+            k: self.get_value(k)
+            for k in self._values.keys()
+            if k not in (TASK_IO_OPENER, TASK_IO_DATASAMPLES, TASK_IO_CHAINKEYS)
+        }
```

Comment:
I think it's not the best option to filter on resource identifiers + introduce the term "dynamic", because:

1. Adding a new term ("dynamic") adds to the cognitive load (more concepts, more complexity). It also means it must be documented.
2. Using hardcoded identifiers carries complexity because both the lib and the client must honor an agreed-upon convention ("the openers must have the identifier 'opener'")

A simpler solution would be to have the client specify the asset "kind" of each resource (in the same way it specifies "multiple")

Then, callers could explicitly filter which kinds they're interested in

```suggestion
    def filter(self, kinds: List[str] = None):
        return {
            k: self.get_value(k)
            for k in self._values.keys()
            if not kinds or v.kind in kinds
        }
        
     def single(self, kind):
       res = [
           self.get_value(k) 
           for (k, v) in self._values.items()
           if v.kind == kind
       ]
       return res[0] if res else None
```

```py
self.task_inputs = {
  k: v
  for k, v in inputs.filter(kinds=[ASSET_KIND_MODEL])
} if inputs else {}

self.task_outputs = {
  k: v 
  for k, v in outputs.filter(kinds=[ASSET_KIND_MODEL, ASSET_KIND_PERFORMANCE])
} if outputs else {}
```

(my filters might be wrong here)

```py
opener_path = inputs.single(kind=ASSET_KIND_OPENER)
```

Reply:
I fully agree with you on this one. The issue is that adding the kind to the cli would have decrease even more the max number of inputs allowed for each task (because of the cli issue) 
Until we do not pass the task inputs and outputs within a file I think we shouldn't go for this option. 
If ok with you I'll add a card in the GT asana board so we keep track of this

Reply:
> The issue is that adding the kind to the cli would have decrease even more the max number of inputs allowed for each task (because of the cli issue)

I understand. This issue could be addressed by passing the algo inputs via a separate `--inputs` parameter instead. This would reduce duplication of information, it seems beneficial regardless of whether we use stdin or a file, and it would allow us to avoid introducing the issues I described above.

Reply:
After discussion with Louis, this will be done in a future PR since this PR is blocking

1. Consider a new parameter `algo-inputs`
2. Within substra-tools, use kind filters instead of identifier filters